warehouse/api/grpc.go

```diff
@@ -1032,3 +1047,109 @@ func (g *GRPC) GetFirstAbortedUploadInContinuousAbortsByDestination(
 
 	return &proto.FirstAbortedUploadInContinuousAbortsByDestinationResponse{Uploads: uploads}, nil
 }
+
+// GetSyncLatency returns the sync latency for the given workspace, destination, start time and aggregation minutes
+// If sourceID is provided, it will return the sync latency for the given sourceID
+// If sourceID is not provided, it will return the sync latency for all sources of the given destination
+func (g *GRPC) GetSyncLatency(ctx context.Context, request *proto.SyncLatencyRequest) (*proto.SyncLatencyResponse, error) {
+	log := g.logger.Withn(
+		obskit.WorkspaceID(request.WorkspaceId),
+		obskit.SourceID(request.SourceId),
+		obskit.DestinationID(request.DestinationId),
+		logger.NewStringField("startTime", request.GetStartTime()),
+		logger.NewStringField("aggregationMinutes", request.GetAggregationMinutes()),
+	)
+	log.Infon("Getting sync latency")
+
+	if request.GetWorkspaceId() == "" || request.GetDestinationId() == "" {
+		return &proto.SyncLatencyResponse{},
+			status.Error(codes.Code(code.Code_INVALID_ARGUMENT), "workspaceID and destinationID cannot be empty")
+	}
+	if request.GetStartTime() == "" {
+		return &proto.SyncLatencyResponse{},
+			status.Errorf(codes.Code(code.Code_INVALID_ARGUMENT), "start time cannot be empty")
+	}
+	if request.GetAggregationMinutes() == "" {
+		return &proto.SyncLatencyResponse{},
+			status.Errorf(codes.Code(code.Code_INVALID_ARGUMENT), "aggregation minutes cannot be empty")
+	}
+
+	startTime, err := time.Parse(time.RFC3339, request.GetStartTime())
+	if err != nil {
+		return &proto.SyncLatencyResponse{},
+			status.Errorf(codes.Code(code.Code_INVALID_ARGUMENT), "start time %s should be in correct %s format", request.GetStartTime(), time.RFC3339)
+	}
+	minStartTime := g.now().AddDate(0, -g.config.maxLatencyQueryLookbackDays, 0)
+	if startTime.Before(minStartTime) {
+		return &proto.SyncLatencyResponse{},
+			status.Errorf(codes.Code(code.Code_INVALID_ARGUMENT), "start time cannot be older than %d days", g.config.maxLatencyQueryLookbackDays)
+	}
+
+	aggregationMinutes, err := strconv.Atoi(request.GetAggregationMinutes())
+	if err != nil {
+		return &proto.SyncLatencyResponse{},
+			status.Errorf(codes.Code(code.Code_INVALID_ARGUMENT), "aggregation minutes %s should be an integer", request.GetAggregationMinutes())
+	}
+
+	srcMap, _ := g.bcManager.ConnectionSourcesMap(request.GetDestinationId())
+	if len(request.GetSourceId()) > 0 {
+		if _, ok := srcMap[request.GetSourceId()]; !ok {
+			return &proto.SyncLatencyResponse{},
+				status.Error(codes.Code(code.Code_UNAUTHENTICATED), "unauthorized request")
```

Comment:
will this translate to 401 in the end?

Reply:
Yes.
```
	// The request does not have valid authentication credentials for the
	// operation.
	//
	// HTTP Mapping: 401 Unauthorized
	Code_UNAUTHENTICATED Code = 16
```

Reply:
why unauthenticated? shouldn't this be some 404 not found error? If we return a 401, webapp tries to do a refresh the access token if I am not wrong.

Reply:
Here we are verifying for sources corresponding to the following destination ID. if that source does not belong to a user we are returning 401 Unauthorized.